vcd/resource_vcd_rde.go

```diff
@@ -0,0 +1,437 @@
+
+	// As RDEs can have many instances with same name and RDE Type, we can't guarantee that we will read the one we want,
+	// but at least we try to filter a bit with things we know, like Organization.
+	var filteredRdes []*govcd.DefinedEntity
+	org, err := vcdClient.GetOrgFromResource(d)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve the Organization of the RDE '%s': %s", name, err)
+	}
+	for _, rde := range rdes {
+		if rde.DefinedEntity.Org != nil && org.Org.ID == rde.DefinedEntity.Org.ID {
+			filteredRdes = append(filteredRdes, rde)
+		}
+	}
+
+	if len(filteredRdes) == 0 {
+		return nil, fmt.Errorf("no RDEs found with name '%s' and RDE Type ID '%s' in Org '%s': %s", name, rdeTypeId, org.Org.Name, govcd.ErrorEntityNotFound)
+	}
+
+	// If there is more than one RDE, we retrieve the IDs to give the user some feedback.
+	var filteredRdesIds []string
+	if len(filteredRdes) > 1 {
+		for _, rde := range filteredRdes {
+			filteredRdesIds = append(filteredRdesIds, rde.DefinedEntity.ID)
+		}
+	}
+
+	err = fmt.Errorf("there are %d RDEs with name '%s' and RDE Type ID '%s' in Org '%s': %v", len(filteredRdes), name, rdeTypeId, org.Org.Name, filteredRdesIds)
+	// We end early with the data source if there is more than one RDE found.
+	if origin == "datasource" && len(filteredRdes) > 1 {
+		return nil, err
+	}
+
+	if len(filteredRdes) > 1 {
+		util.Logger.Printf("[WARN] %s: Choosing the first one", err.Error())
+	}
+	// We just perform another GET by ID to retrieve the ETag of the RDE (it is only returned when we get a specific and unique ID).
+	// We pick the first retrieved RDE from the result above. This doesn't guarantee that is the RDE we want, but it's
+	// how VCD works.
+	return vcdClient.GetRdeById(filteredRdes[0].DefinedEntity.ID)
+}
+
+func resourceVcdRdeUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+	rde, err := getRde(d, vcdClient, "resource")
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	jsonEntity, err := getRdeJson(vcdClient, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = rde.Update(types.DefinedEntity{
+		Name:       d.Get("name").(string),
+		ExternalId: d.Get("external_id").(string),
+		Entity:     jsonEntity,
+	})
+	if err != nil {
+		return diag.Errorf("could not update the Runtime Defined Entity '%s': %s", rde.DefinedEntity.Name, err)
+	}
+
+	if d.Get("resolve").(bool) {
+		err = rde.Resolve()
+		if err != nil {
+			return diag.Errorf("could not resolve the Runtime Defined Entity '%s': %s", rde.DefinedEntity.Name, err)
+		}
+	}
+
+	return resourceVcdRdeRead(ctx, d, meta)
+}
+
+func resourceVcdRdeDelete(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+	rde, err := getRde(d, vcdClient, "resource")
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if d.Get("resolve_on_removal").(bool) {
+		err = rde.Resolve()
+		if err != nil {
+			return diag.FromErr(err)
```

Comment:
Detailed error should aid user IMO

```suggestion
			return diag.Errorf("could not resolve the Runtime Defined Entity before removal '%s': %s", rde.DefinedEntity.Name, err)
```

Reply:
Nice, thanks!